pkg/i2gw/ingress2gateway_test.go

```diff
@@ -150,3 +151,59 @@ func Test_constructIngressesFromCluster(t *testing.T) {
 		})
 	}
 }
+
+func Test_constructProviders(t *testing.T) {
+	testCases := []struct {
+		name           string
+		providers      []string
+		expectingError bool
+	}{{
+		name:           "Test contruct providers with default providers",
+		providers:      GetSupportedProviders(),
+		expectingError: false,
+	}, {
+		name:           "Test contruct providers with provider that not supported",
+		providers:      []string{"ingress-nginx", "fake-provide"},
+		expectingError: true,
+	}}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			cl := fake.NewClientBuilder().WithRuntimeObjects([]runtime.Object{}...).Build()
+			providerByName, err := constructProviders(&ProviderConf{
+				Client: cl,
+			}, tc.providers)
+			if tc.expectingError && err == nil {
+				t.Errorf("Expected error but got none")
+			}
+			if !tc.expectingError {
+				if err != nil {
+					t.Errorf("Expected no error but got %v", err)
+				}
+				if len(tc.providers) != len(providerByName) {
+					t.Errorf("Expected contructed providers num is %d but got %d", len(tc.providers), len(providerByName))
+				}
+				for key := range providerByName {
+					if !slices.Contains(tc.providers, string(key)) {
+						t.Errorf("An unexpected provider %s was constructed", key)
+					}
+				}
```

Comment:
Same as the actual `constructProviders` function, you can take advantage of the fact that `providerByName` is a map, and avoid the usage of `slices.Contains`

Secondly, I think this test would be a bit easier to understand if you added an `expectedProviders` field to each test case.

Reply:
Thanks very much, I will fix these issues immediately.